lisa/wlgen/rta.py

```diff
@@ -534,6 +535,72 @@ def get_cpu_calibrations(cls, target, res_dir=None):
         with cm, target.disable_idle_states():
             return cls._calibrate(target, res_dir)
 
+    @classmethod
+    def _compute_task_map(cls, trace, names):
+        prefix_regexps = {
+            prefix: re.compile(r"^{}(-[0-9]+)*$".format(re.escape(prefix)))
+            for prefix in names
+        }
+
+        task_map = {
+            prefix: sorted(
+                task_id
+                for task_id in trace.task_ids
+                if re.match(regexp, task_id.comm)
+            )
+            for prefix, regexp in prefix_regexps.items()
+        }
+
+        missing = sorted(prefix for prefix, task_ids in task_map.items() if not task_ids)
+        if missing:
+            raise RuntimeError("Missing tasks matching the following rt-app profile names: {}"
+                                .format(', '.join(missing)))
+        return task_map
+
+    # Mapping of Trace objects to their task map.
+    # We don't want to keep traces alive just for this cache, so we use weak
+    # references for the keys.
+    _traces_task_map = weakref.WeakKeyDictionary()
```

Comment:
It feels weird to read this between function definitions. Or my eyes just got tricked and it isn't? Shouldn't it be moved into init()?

Reply:
no, that's a class attribute

Reply:
Defining it between function still is weird. Define it at the top just after declaring the class where all the other attributes are makes more sense, no?

Reply:
The main reason I put it there is because it's right above the only method that uses it, so one does not have to scroll endlessly to see what goes together. It's not something public-facing in any case, so it's not a big deal if people don't notice it at first sight

Reply:
This is deviating from best practices in general IMO. But it's not a big deal if you really prefer it this way. FWIW people don't rely on scrolling when inspecting code only - but could expect all class attributes to be grouped together at the top.

Reply:
My point is that grouping by itself is useless, what matters is what grouping brings. It's useful to group public attributes if someone wants to discover what the class has to offer. Since that attribute is definitely not part of that list, it sounds more useful to have it close to where it's actually used. This attribute is actually an internal persistent state of `resolve_trace_task_names()` and nothing else. If I could, I would declare it inside the function but Python does not have `static` local variables as in C.